### PR TITLE
fix: disable tts when stream goes offline

### PR DESCRIPTION
**Pull request by SputNikPlop**

### CODE REVIEW

File: lib/models/tts.dart [hunk 1]
```diff
@@ -277,6 +277,13 @@ class TtsModel extends ChangeNotifier {
     if (!enabled && !force) {
       return;
     }
+
+    if (model is StreamStateEventModel) {
```
Comment (kevmo314) [inline lib/models/tts.dart:281]: This should be done in the TTS isolate instead since the screen may not be open.

**Review by kevmo314 — CHANGES_REQUESTED** [review]

File: lib/models/tts.dart (hunk 1 above)
Comment (SputNikPlop) [reply lib/models/tts.dart]: Good catch, thought it could be a general tts function.

File: lib/tts_isolate.dart [hunk 2]
```diff
@@ -83,6 +83,16 @@ void vocalizeMessage(Map<String, dynamic>? message) {
     return;
   }
 
+  //check if message model is StreamStateEventModel
+  if (message.containsKey('isOnline')) {
+    bool isOnline = message['isOnline'] as bool;
+
+    if (!isOnline) {
+      TextToSpeechPlugin.stopSpeaking();
```
Comment (kevmo314) [inline lib/tts_isolate.dart:91]: This is on the correct thread now, however it won't disable the icon either. The icon will still show that TTS is enabled. Think about what manages the state of TTS being enabled/disabled? This isolate listens on that state, we need to adjust the button to also listen on that state.

**Review by kevmo314 — CHANGES_REQUESTED** [review]

**Review by jlw4049 — APPROVED** [review]
Looks good, I tested it on both physical/virtual android it is behaving correctly, although not sure how to simulate a stream going offline.

Edit: Doesn't disable notification or tts when stream dies.

File: lib/screens/home.dart [hunk 3]
```diff
@@ -178,6 +181,17 @@ class _HomeScreenState extends State<HomeScreen> with TickerProviderStateMixin {
         model.showAudioPermissionDialog(context);
       }
     });
+    initSharedPreference();
+  }
+
+  initSharedPreference() async {
+    streamPrefs = await StreamingSharedPreferences.instance;
+    streamPrefs.getString('tts_channel', defaultValue: '{}').listen((value) {
```
Comment (kevmo314) [inline lib/screens/home.dart:189]: Add an unsubscribe call in `dispose()`

File: lib/screens/home.dart [hunk 4]
```diff
@@ -178,6 +181,17 @@ class _HomeScreenState extends State<HomeScreen> with TickerProviderStateMixin {
...
+      setState(() {
+        ttsChannel = value;
+        isTTsActive = value.isEmpty || value == "{}" ? false : true;
```
Comment (kevmo314) [inline lib/screens/home.dart:192]: Make this a computed property instead

File: lib/screens/home.dart [hunk 5]
```diff
@@ -237,42 +251,49 @@ class _HomeScreenState extends State<HomeScreen> with TickerProviderStateMixin {
...
+                            ? Icons.record_voice_over
+                            : Icons.voice_over_off),
+                        tooltip: AppLocalizations.of(context)!.textToSpeech,
+                        onPressed: () async {
+                          // Toggle the enabled state
 
-                        final streamPrefs =
-                            await StreamingSharedPreferences.instance;
+                          if (!isTTsActive) {
+                            ttsModel.enabled = true;
```
Comment (kevmo314) [inline lib/screens/home.dart:262]: I don't think this line does anything, does it work with it removed?

File: lib/screens/home.dart [hunk 6]
```diff
@@ -237,42 +251,49 @@ class _HomeScreenState extends State<HomeScreen> with TickerProviderStateMixin {
...
-                              body: null,
-                              locked: true,
-                              autoDismissible: false,
-                            ),
-                          );
-                        }
-                      },
-                    );
+                          if (isTTsActive) {
+                            ttsModel.enabled = false;
```
Comment (kevmo314) [inline lib/screens/home.dart:285]: Same here

**Review by kevmo314 — CHANGES_REQUESTED** [review]
Nice, logic looks good just a few cleanup points.

File: lib/tts_isolate.dart [hunk 7]
```diff
@@ -36,9 +36,21 @@ Future<void> isolateMain(
           .snapshots()
           .listen((latestMessage) async {
         if (latestMessage.docs.isNotEmpty) {
-          final textToSpeak = latestMessage.docs[0]['message'] as String?;
-          if (textToSpeak != null) {
-            await ttsQueue.speak(latestMessage.docs[0].id, textToSpeak);
+          final message = latestMessage.docs[0];
+          if (message.data().containsKey('type')) {
+            final type = message['type'] as String?;
+            if (type == "message") {
```
Comment (kevmo314) [inline lib/tts_isolate.dart:42]: Use a switch here

```
switch (type) {
  case "message": ...
  case "stream.offline": ...
}
```

**Review by kevmo314 — APPROVED** [review]
Nice work, make sure the notifications change gets looped in with @SputNikPlop's work, otherwise this looks great.

**Pull request merged**